meta_client_v2/src/meta_impl.rs

```diff
@@ -0,0 +1,513 @@
+            timeout: ReadableDuration::secs(5),
+            cq_count: 8,
+        }
+    }
+}
+
+struct NodeHeartbeatChannel {
+    heartbeat_sender: ClientDuplexSender<PbNodeHeartbeatRequest>,
+    action_cmd_receiver: Option<ClientDuplexReceiver<PbNodeHeartbeatResponse>>,
+}
+
+struct GrpcClient {
+    client: CeresmetaRpcServiceClient,
+    heartbeat_channel: RwLock<NodeHeartbeatChannel>,
+}
+
+// The handler may contain the Self instance which will cause circular
+// dependency.
+type EventHandlers = Vec<Weak<dyn EventHandler + Send + Sync>>;
+
+struct Inner {
+    meta_config: MetaClientConfig,
+    node_meta_info: NodeMetaInfo,
+
+    grpc_client: RwLock<Option<GrpcClient>>,
+    handlers: RwLock<EventHandlers>,
+}
+
+impl Inner {
+    fn new(meta_config: MetaClientConfig, node_meta_info: NodeMetaInfo) -> Result<Self> {
+        let client = Self {
+            meta_config,
+            node_meta_info,
+            grpc_client: RwLock::new(None),
+            handlers: RwLock::new(EventHandlers::new()),
+        };
+
+        Ok(client)
+    }
+
+    fn request_header(&self) -> RequestHeader {
+        RequestHeader {
+            node: self.node_meta_info.node.to_string(),
+            cluster_name: self.meta_config.cluster_name.clone(),
+        }
+    }
+
+    fn node_meta_info(&self) -> NodeMetaInfo {
+        self.node_meta_info.clone()
+    }
+
+    fn get_cluster_name(&self) -> &str {
+        self.meta_config.cluster_name.as_str()
+    }
+
+    fn connect_grpc_client(&self) -> Result<GrpcClient> {
+        let client = self.build_rpc_client();
+        let (sender, receiver) = client
+            .node_heartbeat_opt(CallOption::default())
+            .context(InitHeartBeatStream)?;
+        Ok(GrpcClient {
+            client,
+            heartbeat_channel: RwLock::new(NodeHeartbeatChannel {
+                heartbeat_sender: sender,
+                action_cmd_receiver: Some(receiver),
+            }),
+        })
+    }
+
+    async fn reconnect_heartbeat_channel(&self) {
+        info!("gRPC reconnect begin");
+
+        loop {
+            match self.connect_grpc_client() {
+                Ok(client) => {
+                    *self.grpc_client.write().await = Some(client);
+
+                    info!("gRPC reconnect succeeds");
+                    return;
+                }
+                Err(e) => {
+                    error!("gRPC reconnect failed, err:{}", e);
+
+                    time::sleep(self.error_wait_lease()).await;
+                }
+            }
+        }
+    }
+
+    fn build_rpc_client(&self) -> CeresmetaRpcServiceClient {
+        let cb = ChannelBuilder::new(Arc::new(Environment::new(self.meta_config.cq_count)));
+        CeresmetaRpcServiceClient::new(cb.connect(&self.meta_config.meta_addr))
+    }
+
+    // TODO: Store the value in field
+    fn error_wait_lease(&self) -> Duration {
+        Duration::from_secs(self.meta_config.lease.as_secs() / 2)
```

Comment:
`Duration` can be divided https://doc.rust-lang.org/std/time/struct.Duration.html#impl-Div%3Cu32%3E